### PR TITLE
useEntityRecord (experimental)

### PR DESCRIPTION
**Pull request by adamziel**
## Description
This PR is a minimal subset of https://github.com/WordPress/gutenberg/pull/38135 focused solely on the `useEntityRecord` hook.

The goal of these new APIs is to lower the barrier of entry for new contributors and make the life of existing contributors easier.

### Example usage:
**Before**
```js
const { page, hasResolved } = useSelect(
  ( select ) => {
	  const selectorArgs = [ 'postType', 'page', pageId ];
	  return {
		  page: select( coreDataStore ).getEntityRecord(
			  ...selectorArgs
		  ),
		  hasResolved: select( coreDataStore ).hasFinishedResolution(
			  'getEntityRecord',
			  selectorArgs
		  ),
	  };
  },
  [ pageId ]
);
```

**After:**
```js
const { record, hasResolved } = __experimentalUseEntityRecord( 'postType', 'page', pageId );
```

See also [how the combination of all proposed hooks makes the navigation block ~200 lines leaner](https://github.com/WordPress/gutenberg/pull/38289).

#### `runIf` option

`runIf` was originally proposed in https://github.com/WordPress/gutenberg/pull/38135 to address the following use case from the navigation block:

```js
const rawNavigationMenu = ref
	? getEntityRecord( ...navigationMenuSingleArgs )
	: null;
```

Thinking more about that, maybe if would be better to have `getEntityRecord()` in a separate, nested component. Either way, this can be a separate discussion in a follow-up PR.

#### [An optional **id** argument that can be provided by an EntityProvider](https://github.com/WordPress/gutenberg/pull/38135#issuecomment-1026608034)

@talldan proposed to make `useEntityRecord` more consistent with `useEntityProp` and have an optional `id` that would be backfilled by `EntityProvider`:

https://github.com/WordPress/gutenberg/blob/8fcd7f2b6eb5bf61b092b25837978ed21aa93d42/packages/core-data/src/entity-provider.js#L96-L98

This looks great! My one concern is "What about passing an empty id to reason about a new entity record?" I propose to hold on for now and discuss what an empty `id` argument should do later on once we're on the mutation hooks.

### Test plan

1. Confirm if the tests are green
2. Confirm if the code makes sense

cc @kevin940726 @talldan @gziolo @draganescu @ellatrix @noisysocks @jsnajdr @getdave @Mamaduka @spencerfinnell

### CODE REVIEW

File: packages/core-data/src/hooks/constants.js
Comment (kevin940726) [inline packages/core-data/src/hooks/constants.js]: I think we can rewrite this file into typescript either using `enums` or `const`:


```ts
export enum Status {
  Idle = 'IDLE',
  Resolving = 'RESOLVING',
  Error = 'ERROR',
  Success = 'SUCCESS',
};
```

File: packages/core-data/src/hooks/use-entity-record.js
Comment (kevin940726) [inline packages/core-data/src/hooks/use-entity-record.js]: Why don't we write this file in typescript instead? Could be easier to read and potentially produce nicer type definitions.

File: packages/core-data/src/hooks/use-entity-record.js
Comment (kevin940726) [inline packages/core-data/src/hooks/use-entity-record.js]: I failed to see the need of introducing this as a value of the returned object. Is it common to check if the record is missing? Based on a quick search in the project, I can only find one occurrence in `template-part`.

I'd argue that the name `isMissing` is also slightly confusing. I'd imagine something like `isEmpty` is more commonly used in this context.

Either way, I think it's a simple case to just let the users do it manually via `hasResolved && !! data`. I don't strongly feel the need to remove it either though :).

File: packages/core-data/src/hooks/use-entity-record.js
Comment (kevin940726) [inline packages/core-data/src/hooks/use-entity-record.js]: I think we need to also expose `error` for error handling. Or if it's handled in `data` then I think it makes sense to separate them into different variables.

File: packages/core-data/src/hooks/use-entity-record.js
Comment (kevin940726) [inline packages/core-data/src/hooks/use-entity-record.js]: I know it's just an example but this code could be improved in performance. Newing a `NumberFormat` instance on every render is costly, we can move it outside the Component to only initialize it once. (Also seems like `currency` is missing in the scope?)

We might want to be careful when writing examples like this since they're often the first piece of code new contributors see and copy-paste from 😆 .

File: packages/core-data/src/hooks/test/use-entity-record.js
Comment (kevin940726) [inline packages/core-data/src/hooks/test/use-entity-record.js]: I'd recommend using `@testing-library/react` for this, but if it works it works 😂 .

**Review by kevin940726 — COMMENTED** [review]
Thank you! Left some comments.

Awesome job as always! 💯💯💯
I'll take a look at other PRs this week 👀 .

File: packages/core-data/src/hooks/use-entity-record.js
Comment (adamziel) [reply packages/core-data/src/hooks/use-entity-record.js]: Somehow I missed the fact that we can now use .ts files – thanks!

File: packages/core-data/src/hooks/use-entity-record.js
Comment (adamziel) [reply packages/core-data/src/hooks/use-entity-record.js]: I think I took that from the navigation block hooks, but I agree, it doesn't need to be here. 👍

File: packages/core-data/src/hooks/use-entity-record.js
Comment (adamziel) [reply packages/core-data/src/hooks/use-entity-record.js]: Do you mean `isError` or error details? I don't think we can access the latter, any ideas?

File: packages/core-data/src/hooks/use-entity-record.js
Comment (adamziel) [reply packages/core-data/src/hooks/use-entity-record.js]: Good spot! I think I'll just update that to use actual entity records instead of made up ones.

File: packages/core-data/package.json
Comment (gziolo) [inline packages/core-data/package.json]: It’s a development dependency that is not necessary for the package to work. It should be listed only in the project’s `package.json` file in `devDependencies`.

File: packages/core-data/src/hooks/use-entity-record.ts
Comment (gziolo) [inline packages/core-data/src/hooks/use-entity-record.ts]: What is the practical usage of `record` vs `editedRecord` when reading the record? Why the hook’s consumer would need both in this high-level API?

File: packages/core-data/src/hooks/test/use-entity-record.js
Comment (gziolo) [inline packages/core-data/src/hooks/test/use-entity-record.js]: The case where `hasEdits` is true is not covered in the tests. I’m mostly curious to better understand the motivation to expose the server state vs local state of the record in the hook.

File: packages/core-data/src/hooks/use-entity-record.ts
Comment (gziolo) [inline packages/core-data/src/hooks/use-entity-record.ts]: This JSDoc needs to be fixed. This hook doesn’t delete anything 😅

File: packages/core-data/src/hooks/use-entity-record.ts
Comment (gziolo) [inline packages/core-data/src/hooks/use-entity-record.ts:41]: This made me think that maybe at this level of abstraction it could better match the UI feedback so the name of the property could be `isLoading` and the status for the enum `Loading`. I don’t feel strongly about it, so you can ignore it if you don’t like the idea.

**Comment by gziolo** [comment]
This is close. I would be happy to see it landed once we finalize all the little things. Thank you for breaking the previous work into smaller pieces that should speed up the process once we land this PR ❤️

File: packages/core-data/src/hooks/use-entity-record.ts
Comment (kevin940726) [reply packages/core-data/src/hooks/use-entity-record.ts:41]: I also like the name `isLoading` more but I guess @adamziel deliberately wants to match the current convention of _resolving_ data 😅.

File: packages/core-data/src/hooks/use-entity-record.js
Comment (kevin940726) [reply packages/core-data/src/hooks/use-entity-record.js]: I mean the error details. If we don't have access to the error then I think we should definitely add it 😆 , not in this PR though.

File: packages/core-data/src/hooks/constants.ts
Comment (kevin940726) [inline packages/core-data/src/hooks/constants.ts:1]: Why do we have to disable this rule, a bug in eslint?

File: packages/core-data/src/hooks/test/use-entity-record.js
Comment (kevin940726) [inline packages/core-data/src/hooks/test/use-entity-record.js]: We don't need to use `act()` with RTL's `render`, it does it for us implicitly.

File: packages/core-data/src/hooks/use-entity-record.ts
Comment (kevin940726) [inline packages/core-data/src/hooks/use-entity-record.ts]: `Object` is not a type that we'd like to use in this context, it behaves similarly to `any` that it just silent the errors without any actual gains.

Ideally, we should already have all the types of core entities available somewhere so that we can do this automatically. Given that it'd be a huge amount of work, we could instead try to use [generics](https://www.typescriptlang.org/docs/handbook/2/generics.html) here to let the users specify the correct types.

```ts
const { record } = useEntityRecord<PageRecord>( 'postType', 'page', id );
// record: PageRecord
```

It's a bit more complicated to get right though, I'm willing to help if you encounter any problems!

File: packages/core-data/src/hooks/use-entity-record.ts
Comment (kevin940726) [inline packages/core-data/src/hooks/use-entity-record.ts:20]: I guess we don't need this either, as this is essentially the same as `!isResolving`.

File: packages/core-data/src/hooks/use-entity-record.ts
Comment (kevin940726) [reply packages/core-data/src/hooks/use-entity-record.ts]: Also, if the data is still resolving, I think the type of this value will be `null`?

File: packages/core-data/src/hooks/test/use-entity-record.js
Comment (adamziel) [reply packages/core-data/src/hooks/test/use-entity-record.js]: You're right, thank you!

File: packages/core-data/src/hooks/use-entity-record.ts
Comment (adamziel) [reply packages/core-data/src/hooks/use-entity-record.ts]: > Object is not a type that we'd like to use in this context, it behaves similarly to any that it just silent the errors without any actual gains.

The gains are that we're able to use TypeScript :-) I went for the minimal typing on purpose for the reasons you have mentioned – there aren't many types defined in `core-data`. Thinking about it more, it shouldn't be a big problem for the JavaScript consumers of the code, and if it is then `any` may be moved to the consumer code. So yes, generics sound like a good approach 👍

File: packages/core-data/src/hooks/use-entity-record.ts
Comment (adamziel) [reply packages/core-data/src/hooks/use-entity-record.ts:20]: Not exactly, `isResolving` and `hasResolved` may both be true when the record is being re-requested. It is how the meta selectors in `@wordpress/data` work so I'd rather not change that logic here. I thought it's confusing and now your comment reassured me, so I'll add that to the docstring.

File: packages/core-data/src/hooks/use-entity-record.ts
Comment (adamziel) [reply packages/core-data/src/hooks/use-entity-record.ts:41]: It is my pet peeve – I'd love to call this `isLoading`! My concern is that we say `resolving` everywhere in `@wordpress/data` and `@wordpress/core-data` and I'd rather stick to that convention than introduce a new term for the same concept. My thinking is that I'd rather always say `resolving` than have moments like _ughh is it loading or resolving here? I never rememeber_.

I even thought about deprecating the word `resolving` for loading, but it makes sense that `resolvers` resolve, not load the data.

File: packages/core-data/src/hooks/use-entity-record.ts
Comment (adamziel) [reply packages/core-data/src/hooks/use-entity-record.ts]: Great spot! Fixed

File: packages/core-data/src/hooks/use-entity-record.js
Comment (adamziel) [reply packages/core-data/src/hooks/use-entity-record.js]: Agreed! It is a quite old problem, too – it would be great to get it fixed:

https://github.com/WordPress/gutenberg/blob/70730e61ba3148ebfa2be125d17c8138ab16392c/packages/core-data/src/resolvers.js#L107-L111

I wonder what is the right level of abstraction here. Core-data could be selfish and introduce an action like `setEntityResolutionError`, but that name is already very telling. Resolution is a `@wordpress/data` concept, so perhaps there could be a meta action available like `failResolution`, similarly to how we have `finishResolution`? 

Taking it even further, we lack try/catch around the resolver call:

https://github.com/WordPress/gutenberg/blob/70730e61ba3148ebfa2be125d17c8138ab16392c/packages/data/src/redux-store/index.js#L414-L428

This sends all exceptions straight into the browser's console. I don't see how that's useful. Perhaps we could wrap it in try/catch and handle exceptions at that level? cc @gziolo @youknowriad

Comment (youknowriad) [reply packages/core-data/src/hooks/use-entity-record.js]: I agree that error handling in core-data is not great at the moment, we need to add the errors to the resolution metadata state, add some selectors for it and yeah seems like a good addition to `useEntityRecord` or `useQuerySelect` as well.

File: packages/core-data/src/hooks/use-entity-record.ts
Comment (adamziel) [reply packages/core-data/src/hooks/use-entity-record.ts]: Oh yes, that makes sense – most of the time `editedRecord` is only needed when there are mutations on the table and I also need access to `editEntityRecord` and perhaps `saveEditedEntityRecord`. I think we can remove it from here just fine 👍

File: packages/core-data/src/hooks/test/use-entity-record.js
Comment (adamziel) [reply packages/core-data/src/hooks/test/use-entity-record.js]: Now that editedRecord is removed, I don't think we need `hasEdits` either – let's keep it for the mutations-related PR

File: packages/core-data/src/hooks/constants.ts
Comment (adamziel) [reply packages/core-data/src/hooks/constants.ts:1]: I think so, it just keeps saying `ESLint: 'Status' is already declared in the upper scope on line 1 column 13.(no-shadow`

There's a related discussion on SO: https://stackoverflow.com/questions/63961803/eslint-says-all-enums-in-typescript-app-are-already-declared-in-the-upper-scope

Perhaps we could disable the no-shadow and enable @typescript-eslint/no-shadow at the global level, but that seems beyond the scope of this PR so I went for eslint-disable for now.

Comment (gziolo) [reply packages/core-data/src/hooks/constants.ts:1]: > Perhaps we could disable the no-shadow and enable @typescript-eslint/no-shadow at the global level, but that seems beyond the scope of this PR so I went for eslint-disable for now.

It isn't the first time we would do use the rule from TS version of ESLint 👍🏻

Comment (adamziel) [reply packages/core-data/src/hooks/constants.ts:1]: Here's a PR to adjust the eslint rules: https://github.com/WordPress/gutenberg/pull/38665

File: packages/core-data/src/hooks/use-entity-record.ts
Comment (adamziel) [reply packages/core-data/src/hooks/use-entity-record.ts]: Hey @kevin940726, you will like this PR: [TypeScript definitions for core-data entity records](https://github.com/WordPress/gutenberg/pull/38666)

**Comment by adamziel** [comment]
This one is ready for another review 🎉

File: packages/core-data/src/hooks/use-entity-record.ts
Comment (jsnajdr) [inline packages/core-data/src/hooks/use-entity-record.ts:82]: Given the `status` is exposed as public API, how does a JavaScript (non-TypeScript) consumer work with it? Are there symbolic constants exportes as public API, too, or is it just a string like `"RESOLVING"`?

Comment (kevin940726) [reply packages/core-data/src/hooks/use-entity-record.ts:82]: I believe it will just compile enums down to an object with string values, so yes `status` should just be a string.

File: packages/core-data/src/hooks/use-entity-record.ts
Comment (kevin940726) [reply packages/core-data/src/hooks/use-entity-record.ts:20]: > `isResolving` and `hasResolved` may both be true when the record is being re-requested.

Huh, interesting! Then what's the difference between `hasResolved` and `!!data`?

File: packages/core-data/src/hooks/use-entity-record.ts
Comment (kevin940726) [inline packages/core-data/src/hooks/use-entity-record.ts]: I believe we have a lint rule somewhere that prefers typescript's type syntax over JSDoc syntax, so this should be an error. Not sure why it isn't working in this file 🤔 . I think it makes sense so that we don't have to declare the same types twice.

File: packages/core-data/src/hooks/use-entity-record.js
Comment (adamziel) [reply packages/core-data/src/hooks/use-entity-record.js]: I started some explorations in https://github.com/WordPress/gutenberg/pull/38669 – let's move the discussion there

File: packages/core-data/src/hooks/use-entity-record.ts
Comment (adamziel) [reply packages/core-data/src/hooks/use-entity-record.ts]: I just adjusted it

File: packages/core-data/src/hooks/use-entity-record.ts
Comment (kevin940726) [reply packages/core-data/src/hooks/use-entity-record.ts:20]: Just checked the implementation in:

https://github.com/WordPress/gutenberg/blob/b51dbd6c3ec18e45b114311764bd1a232858ef7d/packages/data/src/redux-store/metadata/selectors.js#L53-L69

> `isResolving` and `hasResolved` may both be true

I think this is not possible? Or maybe I'm missing something elsewhere 🤔 . I guess something like `isSuccess` and `isError` might be more useful than `hasResolved`. It's also arguably more approachable given that the name directly inherits the `status`.

**Comment by draganescu** [comment]
To my eyes this (the code and the PR itself) is great 🎉

**Comment by adamziel** [comment]
This PR hinges on [useQuerySelect](https://github.com/WordPress/gutenberg/pull/38134) where we're still building consensus around whether it should a public API of `@wordpress/data`. I'd rather not block this PR on that discussion, so I went ahead and moved `useQuerySelect` in here as a private and experimental API. This should allow us to merge it and continue the discussion in the other PR separately.

I'd love to get this one merged as soon as there's an approval 🎉

File: packages/core-data/src/hooks/use-entity-record.ts
Comment (adamziel) [reply packages/core-data/src/hooks/use-entity-record.ts:20]: When I said it's confusing, I didn't realize I am the one who's confused :D I think you're right – `hasResolved` and `isResolving` can't both be true. I'm not sure how I concluded they can – thanks for staying alert!

Let's figure out what do to do next. We have three boolean flags: `hasStarted`, `hasFinished`, and `isResolving`. The following combinations are possible:

* false, false, false – the resolution hasn't started yet, the initial state
* true, false, true – the resolution is in progress
* true, true, false – the resolution has finished

so that's three values to represent. I think we need at least `hasStarted` and `isResolving` then, but `hasFinished` and `isResolving` works just as well. 

> what's the difference between hasResolved and !!data?

The resolved data can be any value at all, including `null`, `undefined`, `false, and so on. We can have `hasResolved === true` and `!! data === false`.

File: packages/core-data/src/hooks/use-entity-record.ts
Comment (getdave) [reply packages/core-data/src/hooks/use-entity-record.ts:41]: IMHO we should stick with existing naming conventions. It will be even more confusing to introduce "loading" at this point.

**Review by getdave — APPROVED** [review]
All the concerns here appear to have been addressed but we don't have a ✅ 

I'm going to approve on that basis along with the fact that this is an experimental hook which gives us some flexibility to merge and iterate.

Thanks for all the work you've put in here @adamziel.

**Comment by Mamaduka** [comment]
Great work, @adamziel 🥇 

Based on the latest commits, it looks like this PR also contains the newly proposed `useSelectQuery` as a private API. Maybe we should've used low-level `useSelect` and updated to `useSelectQuery` after #38134 gets merged?

I don't think this is a blocker, but maybe something we can address as a follow-up.

**Comment by adamziel** [comment]
@Mamaduka good point! I think the discussion in #38134 needs some more time so I brought this in as a private API to unblock this PR. I considered using just `useSelect`, but [useEntityRecords](https://github.com/WordPress/gutenberg/pull/38782) would have to duplicate a lot of that logic so I left it in for now. I'm  open to removing it, but let's take that to #38782 as that's where it really shines.

**Comment by gziolo** [comment]
Awesome work, @adamziel. Keep up PRs coming with the rest of functionality to fully cover CRUD operations for WordPress REST API endpoints 👏

I think that `useSelectQuery ` fits great in the current place as an internal utility because it’s shaped on the specific use case tightly coupled to entities and how they are handled through network requests. Let’s see how it evolves when fetching records and handling mutations grow before we decide to move it to the data package. I’m personally curious to see whether `@wordpress/data` needs this type of API, but we need to refactor first all places that can use the most high-level API exposed in `@wordpress/core-data` with this in other similar PRs.

**Pull request merged**
